### PR TITLE
Beginning unit tests + modularization

### PR DESCRIPTION
**Pull request by jkoppel**
@zenna  Here's what I have in store for this project.

Right now, managing dependencies is quite difficult. I had a couple cases when working on this where changing a namespace would cause things to break in subtle ways (because a different method would be called). Knowing what order to include files in omega.jl is hard, and changing it is dangerous. There's also another thing I find scary: including a package in one file can allow it to be used in another. This is especially true in the tests. I fully expect there are places where a file calls a method despite not importing it.

Luckily, Omega is still small, and so this can be fixed with only medium pain.

I've started separating things into their own module/namespace. Some things will get renamed and moved around if I feel I have a good understanding of what it does and think the current name/location could be improved. This should help crack down a lot on implicit dependencies. I also hope it will improve on compilation times.

As part of this, I'm also trying to set up clear public/private boundaries. Things will get moved around as I try to enhance this boundary. Knowing which things are part of a module's public API is a prerequisite for knowing what tests to write.

One thing I'm struggling with in testing is that it's hard to tell when two methods of the same generic function should be considered as different implementations of the same method (in which case they can be given the same spec and tests), vs. when they should be considered two separate methods that happen to share a name (in which case they should be tested separately).

For method locations: I'm thinking of a discipline referred to in the Haskell world as "no orphaned instances." That means that all methods of a generic function are either declared in the same file as the generic function, or the same file as the datatype. I'm open to feedback on this.

One problem I have with Julia is that I can't make a submodule without including the submodule definition in the parent module. (I.e.: I can't write "module Omega.Proj".) One result of this is that I don't yet have a great way to name the namespace containing a test, but I do find the current approach of naming everything "TestNamespace" satisfactory.

Another thing I'd like to do: stamp out functions that have a name-conflict with the standard library. The motivating example is that the base function of index.jl conflicts with the standard base function, which does something completely different. This caused a confusing error message when I first wrapped index.jl into its own module.  Fortunately, the standard base function is deprecated and is being renamed, so nothing needs to be done for this example.

Please let me know how you feel about the direction I'm taking the codebase, and what recommendations you have.

Sincerely,
Jimmy K

### CODE REVIEW

File: test/tests/lift.jl [hunk 1]
```diff
@@ -5,7 +5,7 @@ struct MyType
   x
 end
 
-Omega.lift(:MyType; n=1, mod=Main)
+Omega.lift(:MyType; n=1, mod=Main) # FIXME: Make this work with TestNamespace
```
Comment (jkoppel) [inline test/tests/lift.jl]: @zenna  I was unable to move this file into its own namespace because of this line. I'd like your help doing so.

**Comment by zenna** [comment]
Hi

A lot of changes.  I will go through them in more detail in a moment.  Let me respond to your prose, first.

> Right now, managing dependencies is quite difficult. I had a couple cases when working on this where changing a namespace would cause things to break in subtle ways (because a different method would be called). Knowing what order to include files in omega.jl is hard, and changing it is dangerous. There's also another thing I find scary: including a package in one file can allow it to be used in another. This is especially true in the tests. I fully expect there are places where a file calls a method despite not importing it.

The primary reason files can't be reordered arbitrary is because Julia requires that types are defined before they are used.  One solution to that is to define all your types up front, but then you lose locality between the types and methods defined on them.  I think you are expecting more from the file system than is there.  `include` is just text substitution, for better or worse.

> I've started separating things into their own module/namespace. Some things will get renamed and moved around if I feel I have a good understanding of what it does and think the current name/location could be improved. This should help crack down a lot on implicit dependencies. I also hope it will improve on compilation times.

I agree with putting `Misc`, and `Tests` into a module.  But not the division of whats in `omega/` into `Proj`, `Space` and `Index`.  I think you are going for a submodule based around each core type.  But that approach doesn't quite fit with multimethods in my opinion, because no type owns a method. I think a better delineation would be into `Core` which includes Omega all the RandVar stuff, lifting, `Disributions` for all the primitive distributions currently in `library/` and `InferenceAlgorithms` for the the stuff in inference algorithms. 

> One thing I'm struggling with in testing is that it's hard to tell when two methods of the same generic function should be considered as different implementations of the same method (in which case they can be given the same spec and tests), vs. when they should be considered two separate methods that happen to share a name (in which case they should be tested separately).

This is a good point.  The latter is sometimes called "punning".  Generally I think it is a bad idea but sometimes it is convenient.  For instance `rand(x,y)` is `rand(cond(x,y))` which is convenient but not particularly consistent with other uses of `rand`.  I could be convinced to remove it.  Julia doesn't yet have well defined interfaces, and so there isn't really a true notion of consistency anyway.

> For method locations: I'm thinking of a discipline referred to in the Haskell world as "no orphaned instances." That means that all methods of a generic function are either declared in the same file as the generic function, or the same file as the datatype. I'm open to feedback on this.

Agree mostly but with exceptions, because (i) there is no canonical generic method in Julia, everything is a multimethod. (ii) methods are usually defined on multiple different types.  In these cases it is clearer to me to have the methods defined conceptually Take the methods I have added to `Base.rand` in `src/rand.jl`.  I think putting these methods in with RandVar or with Omega would be a bad idea.

> Another thing I'd like to do: stamp out functions that have a name-conflict with the standard library. The motivating example is that the base function of index.jl conflicts with the standard base function, which does something completely different. This caused a confusing error message when I first wrapped index.jl into its own module. Fortunately, the standard base function is deprecated and is being renamed, so nothing needs to be done for this example.

Sure.  Are there any other examples?

**Comment by jkoppel** [comment]
We discussed 

> This is a good point. The latter is sometimes called "punning". Generally I think it is a bad idea but sometimes it is convenient. For instance rand(x,y) is rand(cond(x,y)) which is convenient but not particularly consistent with other uses of rand. I could be convinced to remove it. Julia doesn't yet have well defined interfaces, and so there isn't really a true notion of consistency anyway.

I don't have a problem with examples like rand. It's a useful name, and it has a different interface (number of arguments) then the other variant rand(x).

Where it does because a problem is where I can reasonably write code where x is either an A or a B, and foo(x) does something completely different depending on whether x is an A or a B.

But first, there's just the documentation problem: I need to figure out whether I should think of two methods as having a common interface or not.

> Agree mostly but with exceptions, because (i) there is no canonical generic method in Julia, everything is a multimethod. (ii) methods are usually defined on multiple different types. In these cases it is clearer to me to have the methods defined conceptually Take the methods I have added to Base.rand in src/rand.jl. I think putting these methods in with RandVar or with Omega would be a bad idea.

Do you mean src/inference.jl ?

Some of these look like punning/overloading rather than implementing existing signatures. However, this is not the case for the implementation of rand for UTuple{Randvar}.

If we could modify the standard libraries, then it would make sense to put this implementation either in Base.Random or next to the implementation of Vector. We can't, so the best we can do is have them here. People like Don Batory have proposed approaches for this where e.g.: if you import both modules A and B, then you also import a "feature interaction module" AB, which is where you'd put such a thing.

Though, on second glance, putting those alongside RandVar does seem doable.

BTW, you can declare generic functions without providing any methods. The syntax is "function foo end". I expect to start using this.

> Sure. Are there any other examples?

Not yet.

**Comment by jkoppel** [comment]
Wrote some tests for the resolve function.

I'm rather confused about the T param of resolve. It seems to be ignored (I passed in Bool for T with a SimpleOmega{Vector{Int}, Float64},

**Comment by jkoppel** [comment]
I wrote some tests for Space.resolve.

I'm confused about the return type. The return type seems to be "whatever the type that's stored in omega is, regardless of T." But....this is not part of the signature! So it seems to be impossible to write a general signature for resolve without changing Omega{I} to Omega{I,V}.

**Pull request merged**